distributed/shuffle/_rechunk.py

```diff
@@ -216,27 +228,215 @@ def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
     if x.size == 0:
         # Special case for empty array, as the algorithm below does not behave correctly
         return da.empty(x.shape, chunks=chunks, dtype=x.dtype)
+    from dask.array.core import new_da_object
 
-    dsk = {}
     token = tokenize(x, chunks)
-    for ndpartial in _split_partials(x, chunks):
-        if all(slc.stop == slc.start + 1 for slc in ndpartial.new):
-            # Single output chunk
-            dsk.update(partial_concatenate(x, chunks, ndpartial, token))
-        else:
-            dsk.update(partial_rechunk(x, chunks, ndpartial, token))
-    layer = MaterializedLayer(dsk)
     name = rechunk_name(token)
-    graph = HighLevelGraph.from_collections(name, layer, dependencies=[x])
-    arr = da.Array(graph, name, chunks, meta=x)
-    return arr
+    disk: bool = dask.config.get("distributed.p2p.disk")
+
+    layer = P2PRechunkLayer(
+        name=name,
+        token=token,
+        chunks=chunks,
+        chunks_input=x.chunks,
+        name_input=x.name,
+        disk=disk,
+    )
+    return new_da_object(
+        HighLevelGraph.from_collections(name, layer, [x]),
+        name,
+        chunks,
+        meta=x._meta,
+        dtype=x.dtype,
+    )
+
+
+class P2PRechunkLayer(Layer):
+    name: str
+    token: str
+    chunks: ChunkedAxes
+    chunks_input: ChunkedAxes
+    name_input: str
+    disk: bool
+    keepmap: np.ndarray
+
+    _cached_dict: _T_LowLevelGraph | None
```

Comment:
I find the pattern where we declare and initialize this attribute at runtime odd (e.g. here: https://github.com/hendrikmakait/distributed/blob/d816b8a6495f26bc7e9e3e1b33e0281aebeaf3c2/distributed/shuffle/_shuffle.py#L214-L217). Is there a good reason for doing so that I miss, maybe some (de-)serialization magic?

Reply:
afaik there is no reason for this